compiler/vm/vmdef.nim

```diff
@@ -90,27 +91,224 @@ type
+  # is expected/used 
+
+  # XXX: once first class openArray support is available during for bootstrapping,
+  # `MemoryHandle` could be turned into `openArray[byte]`
+  MemoryHandle* = object
+    # XXX: without this dummy pointer, memory gets corrupted in ways
+    # that lead to random illegal storage accesses. The corruption very likely
+    # still happens but it seems to be more hidden. It's likely that some part
+    # of the VM writes past their the allocated memory range
+    _: pointer
+    p: ptr Atom
+    l: int
+  VmMemoryRange* = openArray[byte] # TODO: rename to `VmMemoryRegion`
+
+  TypedHandle* = tuple[h: MemoryHandle, typ: PVmType]
+
+  VmSeq* = object
+    length*: int
+    data*: MemoryHandle
+
+  VmString* = distinct VmSeq
+
+  VmSet* = object
+    # XXX maybe use `ptr IntSet`? Can't use IntSet with automatic storage duration since
+    # sizeof(IntSet) is also huge (smaller than set[uint16] however)
+
+    # XXX should `setElems` be ordered? Would speed up contains checks but require sorting on modification
+    # XXX `setElem` could also be run-length encoded for speed and memory efficiency...
+
+    # XXX In the end, VmSets should be embedded into their surrounding storage, without any heap allocations
+    elems* : seq[uint16]
+
+  Atom* {.union.} = object
+    intVal*: BiggestInt ## akInt
+    floatVal*: BiggestFloat ## akFloat
+    ptrVal*: ptr Atom ## akPtr
+    setVal*: VmSet ## akSet
+    strVal*: VmString ## akString
+    seqVal*: VmSeq ## akSeq
+    refVal*: GlobalRef ## akRef
+    callableVal*: VmFunctionObject ## akCallable
+
+    nodeVal*: PNode ## akPNode
+
+  # TODO: needs a better name
+  VmObjectAtom* = object ## A container that stores a value. Single atom
+                         ## values are stored in-place while complex values
+                         ## are seperately allocated 
+    case kind*: AtomKind
+    of akInt..pred(akObject):
+      atom: array[32, byte]
+    of akObject, akArray:
+      obj*: MemoryHandle
+    
+    typ*: PVmType
+
+  
+  # XXX: VmObjectAtom doesn't store pseudo atoms in-place, due to their dynamic size.
+  # However, `TypedVmObject`s are only meant to be used for globals and constants. For these, 
+  # the pseudo-atom flat byte size is both constant and known at initialization time. Thus, 
+  # we could allocate the raw atom or object (pseudo atom) in the same memory location. We'd need a custom
+  # allocator for `TypedVmValue` then, but globals and constants probably need one anyway...
+  TypedVmValue* = VmObjectAtom
+
+  TFullReg* = object
     case kind*: TRegisterKind
-    of rkNone: nil
+    of rkNone: discard
     of rkInt: intVal*: BiggestInt
     of rkFloat: floatVal*: BiggestFloat
-    of rkNode: node*: PNode
-    of rkRegisterAddr: regAddr*: ptr TFullReg
-    of rkNodeAddr: nodeAddr*: ptr PNode
-
+    of rkAddress:
+      addrVal*: MemoryHandle
+      addrTyp*: PVmType
+      isValid*: bool ## Indicates if `addrVal` points to a valid location
```

Comment:
What is the purpose of this field? From my skimming it seems to be an optimization to avoid location validation? If that's the case, just validate always. A mispredicted branch due to this `bool` costs around 20-30 cycles, and if validation is cheaper than that then there is no reason not to always validate instead, especially since there are no indications that it will primarily lean to any one branch during execution.

Reply:
But validation has a high cost? First it has to be checked if the pointer lies in a guest allocated memory region, then if it points to valid location and last if the pointers type matches with the location's. Doing that on every pointer dereference seems very inefficient.

Caching the valid state means that validation only has to happen when trying to dereference an unknown pointer (or after the memory region the pointer pointed to gets released)

Reply:
An another problem with this is that this cached state is an another source-of-truth, and it has higher priority than anything else.

This become trouble when it's one of the main pillars that maintains memory safety for the VM.

> But validation has a high cost? First it has to be checked if the pointer lies in a guest allocated memory region, then if it points to valid location and last if the pointers type matches with the location's. Doing that on every pointer dereference seems very inefficient.

That means optimization has to be put into verification, or to verify it at compile-time and emit special bytecode. Any less is troubling for memory safety.